mapflow/functional/layer_utils.py

```diff
@@ -280,43 +281,35 @@ def add_layer(self, layer: Optional[QgsMapLayer], order=0) -> None:
         else:  # assume user opted to not use a group, add layers as usual
             self.project.addMapLayer(layer)
 
-    def add_preview_layer(self, preview_layer, preview_urls, preview_ids):   
-        id = preview_layer.id()
-        url = preview_layer.dataProvider().dataSourceUri()
-        preview_ids.append(id)
-        # If new preview uri is in a list of previously added previes
-        if url in preview_urls:
-            # Delete it from the list and from the layer tree   
-            preview_urls.remove(url)
-            for l in self.iface.mapCanvas().layers():
-                if url == l.dataProvider().dataSourceUri() and l.id() in preview_ids:
-                    QgsProject.instance().removeMapLayers([l.id()])
-        # If preview list is empty
-        if len(preview_urls) == 0:
-            # Add preview to the bottom and its uri to the list
+    def add_preview_layer(self, preview_layer, preview_dict): 
+        layers_to_delete = []
+        # In a dictionary of ids (as keys) and urls (as values) of plugin added previews
+        for id, url in preview_dict.copy().items():
```

Comment:
Wow, I have an idea. We can do {url:id} dict instead of this.
First, let's clean the preview list from removed layers:
```
for url, id in preview_dict.items():
    if <layer with id is not in project>:
        del preview_dict[url]
    else: 
        <memorize the layer's position and find the top one>

self.add_layer(layer = preview_layer, order = index)
```
So this sections transforms to:
```
url = preview_layer.dataProvider().dataSourceUri()
if url in preview_dict.keys():
    current_preview_id = preview_dict[url]
    # we can't have many layers with the same ID 
    QgsProject.instance().removeMapLayers(current_preview_id)
preview_dict[url]=  preview_layer.id()
```

And then we search for all the layers' positions in the layer view to our data